diff --git a/immut/src/produce.luau b/immut/src/produce.luau
index cab3380..ea1253e 100644
--- a/immut/src/produce.luau
+++ b/immut/src/produce.luau
@@ -59,16 +59,16 @@ type Recipe<T> = (draft: T) -> (T? | typeof(None))
 	end)
 	```
 ]=]
-local function produce<T>(base: T, recipe: Recipe<T>): T?
+local function produce<T>(base: T, recipe: Recipe<T>): T
 	local proxy = isDraftable(base) and Draft.new(base) or base
 
 	local nextValue = recipe(proxy :: T) or proxy
 
 	if nextValue == None then
-		return nil
+		return base
 	end
 
-	return finishDraft(nextValue)
+	return finishDraft(nextValue) :: T
 end
 
 return produce
