diff --git a/node_modules/@rbxts/lemon-signal/dist/index.d.ts b/node_modules/@rbxts/lemon-signal/dist/index.d.ts
index 9830398..128aa17 100644
--- a/node_modules/@rbxts/lemon-signal/dist/index.d.ts
+++ b/node_modules/@rbxts/lemon-signal/dist/index.d.ts
@@ -5,7 +5,7 @@ export declare type SignalParameters<T> = Parameters<
 export declare type SignalCallback<T> = (...args: SignalParameters<T>) => unknown;
 export declare type SignalWait<T> = T extends unknown[] ? LuaTuple<T> : T;
 
-export declare class Connection<T> {
+export declare class Connection {
 	public readonly Connected: boolean;
 
 	public Destroy(): void;
@@ -13,7 +13,7 @@ export declare class Connection<T> {
 	public Reconnect(): void;
 }
 
-export declare class Signal<T> {
+export declare class Signal<T = []> {
 	public readonly RBXScriptConnection?: RBXScriptConnection;
 	public readonly TotalConnections?: number;
 	public readonly OnConnectionsChanged: Signal<number>;
@@ -25,8 +25,8 @@ export declare class Signal<T> {
 		trackConnections?: boolean,
 	) => Signal<Parameters<T>>;
 
-	public Connect(fn: SignalCallback<T>): Connection<T>;
-	public Once(fn: SignalCallback<T>): Connection<T>;
+	public Connect(fn: SignalCallback<T>): Connection;
+	public Once(fn: SignalCallback<T>): Connection;
 	public Wait(): SignalWait<T>;
 	public Fire(...args: SignalParameters<T>): void;
 	public DisconnectAll(): void;
